src/eigensolver/tridiag_solver/misc_gpu_kernels.cu

```diff
@@ -0,0 +1,241 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include "dlaf/eigensolver/tridiag_solver/misc_gpu_kernels.h"
+#include "dlaf/gpu/api.h"
+#include "dlaf/util_cuda.h"
+#include "dlaf/util_math.h"
+
+#include <cuComplex.h>
+#include <thrust/execution_policy.h>
+#include <thrust/merge.h>
+
+namespace dlaf::eigensolver::internal {
+
+// https://github.com/NVIDIA/thrust/issues/1515
+//
+template <class T>
+void mergeIndicesOnDevice(const SizeType* begin_ptr, const SizeType* split_ptr, const SizeType* end_ptr,
+                          SizeType* out_ptr, const T* v_ptr, cudaStream_t stream) {
+  auto cmp = [v_ptr] __device__(const SizeType& i1, const SizeType& i2) {
+    return v_ptr[i1] < v_ptr[i2];
+  };
+  // NOTE: The call may be synchronous, to avoid that either wrap in a __global__ function as shown in
```

Comment:
Is the idea with this that `thrust::merge` has a device-callable implementation as well, so that gives you control of the launch and synchronization instead?

I'd say we go with the current implementation even it may be blocking, open an issue about updating it, profile it and fix it as needed?

Reply:
@teonnik provided more details, and more details here: https://github.com/NVIDIA/thrust/blob/main/examples/cuda/async_reduce.cu. This comment was only to increase my understanding of the potential workaround and does not block this PR.

Reply:
I opened an issue: https://github.com/eth-cscs/DLA-Future/issues/636.

The comment is inspired by an example in Thrust : thrust/examples/cuda/async_reduce.h .However, the approach of wrapping the kernel in a `__global__` function may now be outdated as it appears to rely on CUDA dynamic parallelism. This has to be investigated further.